srml/support/src/traits.rs

```diff
@@ -640,3 +640,46 @@ pub trait ChangeMembers<AccountId> {
 impl<T> ChangeMembers<T> for () {
 	fn change_members(_incoming: &[T], _outgoing: &[T], _new_set: &[T]) {}
 }
+
+/// A generic trait for reporting slashing violations.
+pub trait ReportSlash<Misbehaved, Reporters, Hash, Timestamp> {
+	/// Reports slashing for a misconduct where `to_slash` are the misbehaved entities and
+	/// `to_reward` are the entities that detected and reported the misbehavior
+	///
+	/// Returns `Ok` if the misconduct was unique otherwise `Err`
+	fn slash(to_slash: Misbehaved, to_reward: Reporters, footprint: Hash, timestamp: Timestamp) -> Result<(), ()>;
+}
+
+/// A generic trait for enacting slashes
+pub trait DoSlash<Misbehaved, Severity, Kind, Timestamp> {
+	/// The slashed entries
+	type SlashedEntries;
+
+	/// The total slashed amount
+	type SlashedAmount: Copy + Clone + Codec + SimpleArithmetic;
+
+	/// Performs the actual slashing based on severity
+	///
+	/// Return the slashes entities which may not be the same as `to_slash`
+	/// because history slashes of the same kind may be included
+	fn do_slash(to_slash: Misbehaved, severity: Severity, kind: Kind, timestamp: Timestamp)
+		-> Result<(Self::SlashedEntries, Self::SlashedAmount), ()>;
+}
+
+/// A generic trait for paying out rewards
```

Comment:
it's only for paying out rewards to reporters. not for paying block rewards, finalisation rewards, parachain rewards, ...